agent/tool-scripts/postprocess/pidstat-postprocess

```diff
@@ -70,88 +70,82 @@ if ( (($? >> 8) == 0) && (! -e "/usr/bin/vdsClient") && (-e "/usr/bin/virsh") )
-		my $kb_ccrw = shift(@stats);
-		my $io_delay = shift(@stats);
-		my $cswitches = shift(@stats);
-		my $nvcswitches = shift(@stats);
-		my $cmd = join(' ', @stats);
-
-		# print "timestamp[$timestamp]  uid[$uid]\n";
-
-		my $timestamp_ms = 1000 * $timestamp;
-
-		# if threads are reported
-		if ( $pid =~ /(\d+)\s+(\d+)/ ) {
-			$pid = $1;
-			$tid = $2;
-			if ($tid == 0) {
-				next;
-			} else {
-				$pid = $tid;
-				$cmd =~ s/\|_+//;
-			}
-		} else {
-			$pid =~ s/\s+//g;
-		}
-
-		# if this is a VM, change the "$cmd" to something more readable
-		if ( $cmd =~ /name\s(\S+).*/ ) {
-			$vm_name = $1;
-			$cmd = "KVMguest--" . $vm_name;
-		}
-		$cmd = substr($cmd, 0, 40);
-		$cmd =~ s/\s/_/g;
-		$cmd =~ s/\'//g;
-		$cmd =~ s/\"//g;
-                # commas cause problems in the CSV files.
-                $cmd =~ s/,//g;
-		# avoid 'Invalid conversion in printf: "%=" ' errors if cmd contains a %.
-		$cmd =~ s/%/%%/g;
-		$pid = $pid . "-". $cmd;
-		$pidstat{cpu_usage}{percent_cpu}{$pid}{$timestamp_ms} = $cpu;
-		$pidstat{file_io}{io_reads_KB_sec}{$pid}{$timestamp_ms} = $kb_read;
-		$pidstat{file_io}{io_writes_KB_sec}{$pid}{$timestamp_ms} = $kb_write;
-		$pidstat{memory_faults}{minor_faults_sec}{$pid}{$timestamp_ms} = $minor_faults;
-		$pidstat{memory_faults}{major_faults_sec}{$pid}{$timestamp_ms} = $major_faults;
-		$pidstat{context_switches}{voluntary_switches_sec}{$pid}{$timestamp_ms} = $cswitches;
-		$pidstat{context_switches}{nonvoluntary_switches_sec}{$pid}{$timestamp_ms} = $nvcswitches;
-		$pidstat{memory_usage}{virtual_size}{$pid}{$timestamp_ms} = $vsz;
-		$pidstat{memory_usage}{resident_set_size}{$pid}{$timestamp_ms} = $rss;
-		$timestamps{$timestamp_ms}++;
+my %timestamps;
+my @attributes = get_pidstat_attributes();
+my $pids_dir = $dir . "/pids";
+if (opendir(my $pids_dh, $pids_dir) || die "Could not open $pids_dir") {
+        for my $pid (grep(/\d+/, readdir($pids_dh))) {
+		my $pid_dir = $pids_dir . "/" . $pid;
+		opendir(my $pid_dh, $pid_dir) || die "Could not open $pid_dir";
+		# A PID may be recycled over time as 1 process exits and another later is created,
+		# and therfore more than 1 process could have the same PID
+		my $timestamp_ms;
+        	for my $filename (grep(/[^.]/, readdir($pid_dh))) {
+			my $cmd;
+			my $pid_cmd;
+			open(my $fh, $pid_dir . "/" . $filename) || die "Could not open $filename";
+			my %old_stats;
+			my $line_num = 0;
+			while ($line = <$fh>) {
+				$line_num++;
+				if ($line_num == 1) {
+					# get the real, unmodified cmdline first
+					$cmd = $line;
+					chomp $cmd;
+					# convert qemu to a more readable label with VM name
+					if ( $cmd =~ /qemu/ and $cmd =~ /name\s(\S+).*/ ) {
+						$vm_name = $1;
+						$cmd = "KVMguest--" . $vm_name;
+					}
+					# modify to be appropriate as a label for a series in a graph
+					$cmd =~ s/%/%%/g; # avoid 'Invalid conversion in printf: "%=" ' errors if cmd contains a %.;
+					$cmd =~ s/[\'|\"]//g; # quotes cause problems everywhere
```

Comment:
This gets rid of '|' chars as well: is that intended?

Reply:
No, it was not.

Reply:
But it seems like it is a good thing none-the-less.  So let's leave it there.

Reply:
In that case, the comment should be fixed.